### PR TITLE
boats: Improve physics of boat by implementing water Drag and Friction/Viscosity Force

### PR DESCRIPTION
**Pull request by gang65**
I'm a Physicist, and I love Minetest.
I'm interesting in improving physics of Minetest. I started with the Boat.

With this commit I improve drag force of boat, according to equation:
  Drag_force = Drag_coefficient * Speed ^2
  

More information about drag force is available at:
  https://en.wikipedia.org/wiki/Drag_(physics)

By adding small constant force I also implementing following forces: 
 https://en.wikipedia.org/wiki/Friction#Fluid_friction
 https://en.wikipedia.org/wiki/Viscosity

After that implementation, the boat is moves more realistic.

### CODE REVIEW

File: mods/boats/init.lua [hunk 1]
```diff
@@ -181,15 +181,12 @@ function boat.on_step(self, dtime)
...
-	self.v = self.v - dtime * 0.6 * s
+	self.v = self.v - dtime * 0.2 * self.v * self.v
 	if s ~= get_sign(self.v) then
 		self.object:set_velocity({x = 0, y = 0, z = 0})
 		self.v = 0
 		return
 	end
-	if math.abs(self.v) > 5 then
-		self.v = 5 * get_sign(self.v)
-	end
```
Comment (gang65) [inline mods/boats/init.lua]: We don't need maximum value anymore, as Drag Force is dependent from Speed^2.
It means that it the boat speed is already limited.

**Comment by gang65** [comment]
@paramat Do you know why tests are failing. I didn't touch this part of code

File: mods/boats/init.lua [hunk 2]
```diff
@@ -181,15 +181,12 @@ function boat.on_step(self, dtime)
 		return
 	end
 	local s = get_sign(self.v)
-	self.v = self.v - dtime * 0.6 * s
+	self.v = self.v - s * dtime * 0.2 * self.v * self.v
```
Comment (DS-Minetest) [inline mods/boats/init.lua]: This currently changes the maximum velocity.
```
a = 4.0
v_max = 5
c * v_max * v_max = a
c = a / v_max / v_max = 4/25 = 0.16 ~= 0.2
```
I suggest to use `0.16` instead of `0.2` or/and to change the boats acceleration.

Comment (gang65) [reply mods/boats/init.lua]: Nice catch ! Thanks.

I have increased the boat acceleration from 4.0 to 5.0.

I have also optimized code a little bit. Now it is much faster than previously.
The code could be optimized further, but I would like to leave it for next Pull Request.

**Comment by paramat** [comment]
Don't worry, if you look at the failing test it is in another mod, it will get fixed independently.
Nice boat feature, i did something similar in my driftcar mod.

File: mods/boats/init.lua [hunk 3]
```diff
@@ -180,15 +180,13 @@ function boat.on_step(self, dtime)
 		self.object:set_pos(self.object:get_pos())
 		return
 	end
-	local s = get_sign(self.v)
-	self.v = self.v - dtime * 0.6 * s
-	if s ~= get_sign(self.v) then
-		self.object:set_velocity({x = 0, y = 0, z = 0})
+	-- We need to multiple by abs to not loose sign of velocity
```
Comment (pgimeno) [inline mods/boats/init.lua]: loose -> lose

**Comment by SmallJoker** [comment]
Indeed, the movement is now much better. However, I think that the boat now speeds up way too fast. Would it be possible to lower the acceleration?

File: mods/boats/init.lua (hunk 3 above)
Comment (gang65) [reply mods/boats/init.lua]: Fixed

**Comment by gang65** [comment]
@SmallJoker I have decreased the acceleration. It is now similar to previous implementation (of course it will be always different as the acceleration is not linear now), with preservation of the maximum speed.

**Comment by gang65** [comment]
I would like to implement drifting and implement proper buoyancy.
I plan to create next Merge Request after this will be accepted (to avoid conflicts).

What should I do next, to make this Pull Request accepted and merged into master?

**Comment by paramat** [comment]
What do you mean by drifting?

This PR may be fine now, we just have very limited dev time and this is not urgent, so it may take a while before this gets merged. I'll try to test it as soon as i can.

File: mods/boats/init.lua [hunk 4]
```diff
@@ -180,15 +180,13 @@ function boat.on_step(self, dtime)
 		self.object:set_pos(self.object:get_pos())
 		return
 	end
-	local s = get_sign(self.v)
-	self.v = self.v - dtime * 0.6 * s
-	if s ~= get_sign(self.v) then
-		self.object:set_velocity({x = 0, y = 0, z = 0})
+	-- We need to multiple by abs to not lose sign of velocity
+	local drag = dtime * 0.08 * self.v * math.abs(self.v)
+	-- If drag is larger than velocity, then stop horizontal move.
```
Comment (paramat) [inline mods/boats/init.lua]: "then stop horizontal movement" with no full stop.

**Comment by sofar** [comment]
The movement is better but "over the top" when it comes to stopping: it's virtually impossible to come to a reasonably "close to" full stop somehow now, since even the slightest movement will forever and ever just make the boat slowly move slower and slower.

I'd like to see maybe some residual friction force apply at all times:

```
diff --git a/mods/boats/init.lua b/mods/boats/init.lua
index c6100182..ceb1a97a 100644
--- a/mods/boats/init.lua
+++ b/mods/boats/init.lua
@@ -181,7 +181,7 @@ function boat.on_step(self, dtime)
                return
        end
        -- We need to multiple by abs to not lose sign of velocity
-       local drag = dtime * 0.08 * self.v * math.abs(self.v)
+       local drag = dtime * 0.08 * self.v * math.abs(self.v) + 0.01
        -- If drag is larger than velocity, then stop horizontal move.
        if math.abs(self.v) <= math.abs(drag) then
                self.v = 0
```

This way, it's relatively easy to get the boat to a full stop. You can still glide a long time.

**Comment by sofar** [comment]
I would support `drifting` but please note the boat is not a `sphere` but has clear straight sides. This wouldn't make the boat drift entirely at all, it would still have a significant rudder effect, Also note the boat has an actual rudder (that we could animate!) that further would suggest that the boat may take corners much smoother than it does right now, but it wouldn't actually drift much like a car on ice at all.

File: mods/boats/init.lua (hunk 4 above)
Comment (gang65) [reply mods/boats/init.lua]: Updated

**Comment by gang65** [comment]
@sofar Thanks for your suggestion with friction force. 
I have implemented that, and it is working much better now.

File: mods/boats/init.lua [hunk 5]
```diff
@@ -9,10 +9,12 @@ end
...
-		return 0
+	if i > 0 then
+		return 1
+	elseif i < 0 then
+		return -1
 	else
-		return i / math.abs(i)
+		return 0
 	end
 end
```
Comment (DS-Minetest) [inline mods/boats/init.lua]: Changing something here seems a bit out of scope.

Btw. there's also a `math.sign` from builtin which does the same:
https://github.com/minetest/minetest/blob/42e1a127140965cac1be6e51e48192e341c2a29e/doc/lua_api.txt#L2437-L2441

Ergo `local get_sign = math.sign` would be enough.

**Comment by paramat** [comment]
Ah that kind of drifting. Yes, since water is so dense and boat sides straight (i want to replace this ugly raft with more of a boat shape) then i think drifting will be so minimal it won't be worth simulating.
I don't want boats to get too complex in physics modelling. Buoyancy will be worth working on though.

File: mods/boats/license.txt [hunk 6]
```diff
@@ -33,6 +33,7 @@ Copyright (C) 2012-2016 Zeg9
 Copyright (C) 2012-2016 thetoon
 Copyright (C) 2012-2016 PavelS(SokolovPavel)
 Copyright (C) 2016 sofar (sofar@foo-projects.org)
+Copyright (C) 2019 Bartosz (gang65@poczta.onet.pl)
```
Comment (paramat) [inline mods/boats/license.txt]: These are credits for media not code, so you shouldn't add yourself.

**Comment by pgimeno** [comment]
I wish something like this was done for the player as well.

**Comment by paramat** [comment]
Lua warnings:
```
Checking mods/boats/init.lua                      2 warnings
    mods/boats/init.lua:135:47: accessing undefined field 'sign' of global 'math'
    mods/boats/init.lua:175:23: accessing undefined field 'sign' of global 'math'
```

**Comment by gang65** [comment]
> Lua warnings:
> 
> ```
> Checking mods/boats/init.lua                      2 warnings
>     mods/boats/init.lua:135:47: accessing undefined field 'sign' of global 'math'
>     mods/boats/init.lua:175:23: accessing undefined field 'sign' of global 'math'
> ```

Do you know how to fix that?
These functions is defined at:
https://github.com/minetest/minetest/blob/master/builtin/common/misc_helpers.lua#L236

As I noticed, it is imported correctly.
Should I add some `include` or something?

**Comment by paramat** [comment]
I see, it's a failure of luacheck, not sure what to do yet so leave it as it is, until we get opinion from other core devs.

**Comment by sofar** [comment]
Minetest modifies `math` to include these extra functions, but luacheck assumes that is not done, therefore, it warns. Luacheck has no way of knowing about this, so it's an absolute false positive. The same happens for `table.copy` which is also not standard, and many mods patch their `.luackeckrc` for it.

**Comment by gang65** [comment]
@sofar Do you know where I could find some example patch for `.luackeckrc`?

**Comment by paramat** [comment]
See https://github.com/minetest/minetest_game/blob/0ea6065a09f1fee014c7f2a9a36605d57986a209/.luacheckrc#L14

File: mods/boats/license.txt (hunk 6 above)
Comment (paramat) [reply mods/boats/license.txt]: Resolved.

**Comment by gang65** [comment]
Hello.
Do you know what needs to be done, to accept this Pull Request?

File: mods/boats/init.lua [hunk 7]
```diff
@@ -180,15 +171,13 @@ function boat.on_step(self, dtime)
...
-	if s ~= get_sign(self.v) then
-		self.object:set_velocity({x = 0, y = 0, z = 0})
+	-- We need to preserve velocity sign to properly apply drag force while moving backward
+	local drag = dtime * math.sign(self.v) * (0.01 + 0.08 * self.v * self.v)
+	-- If drag is larger than velocity, then stop horizontal movement.
+	if math.abs(self.v) <= math.abs(drag) then
 		self.v = 0
-		return
-	end
-	if math.abs(self.v) > 5 then
```
Comment (gang65) [inline mods/boats/init.lua]: There is no need to artificially limit the maximum speed of the boat. It is limited by the drag force.

**Comment by paramat** [comment]
It is not the code that is delaying this, it is lack of core dev time.
So ... if you paid us all full time wages ... =)
The concept of this is good.

File: mods/boats/init.lua (hunk 7 above)
Comment (paramat) [reply mods/boats/init.lua]: Correct =)

File: mods/boats/init.lua [hunk 8]
```diff
@@ -180,15 +171,13 @@ function boat.on_step(self, dtime)
 		self.object:set_pos(self.object:get_pos())
 		return
 	end
-	local s = get_sign(self.v)
-	self.v = self.v - dtime * 0.6 * s
-	if s ~= get_sign(self.v) then
-		self.object:set_velocity({x = 0, y = 0, z = 0})
+	-- We need to preserve velocity sign to properly apply drag force while moving backward
+	local drag = dtime * math.sign(self.v) * (0.01 + 0.08 * self.v * self.v)
```
Comment (paramat) [inline mods/boats/init.lua]: The 0.08 needs altering slightly so that max velocity remains exactly 5n/s.
Seems to need to be:
1.99/25 = 0.0796

**Comment by paramat** [comment]
I am keen to merge this now so will probably +1 once my line comment is attended too.

File: mods/boats/init.lua [hunk 9]
```diff
@@ -180,15 +171,13 @@ function boat.on_step(self, dtime)
 		self.object:set_pos(self.object:get_pos())
 		return
 	end
-	local s = get_sign(self.v)
-	self.v = self.v - dtime * 0.6 * s
-	if s ~= get_sign(self.v) then
-		self.object:set_velocity({x = 0, y = 0, z = 0})
+	-- We need to preserve velocity sign to properly apply drag force while moving backward
```
Comment (paramat) [inline mods/boats/init.lua]: This line is just over 90 columns (tab size 4), please wrap.

File: mods/boats/init.lua [hunk 10]
```diff
@@ -180,15 +171,13 @@ function boat.on_step(self, dtime)
...
+	local drag = dtime * math.sign(self.v) * (0.01 + 0.08 * self.v * self.v)
+	-- If drag is larger than velocity, then stop horizontal movement.
```
Comment (paramat) [inline mods/boats/init.lua]: Single line comment so no full stop needed (see other comments)

**Comment by paramat** [comment]
Code looks fine apart from line comments.
Tested and works well.
+1 once line comments attended to.

File: mods/boats/init.lua (hunk 9 above)
Comment (gang65) [reply mods/boats/init.lua]: Done

File: mods/boats/init.lua (hunk 8 above)
Comment (gang65) [reply mods/boats/init.lua]: Updated

File: mods/boats/init.lua (hunk 10 above)
Comment (gang65) [reply mods/boats/init.lua]: Could you please explain what I need to change in this comment?
I think the comment is correct (we need to set velocity to 0, to avoid negative values caused by drag).

Comment (paramat) [reply mods/boats/init.lua]: Sorry if i was unclear, all i request is removing the '.' at the end, single-line comments do not require them, see the other single-line comments in the file.

Comment (paramat) [reply mods/boats/init.lua]: In English the '.' character is sometimes called a 'full stop', maybe this caused confusion =D

Comment (paramat) [reply mods/boats/init.lua]: Actually nevermind this, it is only a tiny change, i will merge this now.

**Comment by paramat** [comment]
:+1:
